erasure_test.go

```diff
@@ -0,0 +1,133 @@
+}
+
+// Generates good encoded data with one parity block and data block missing.
+func genGoodEncodedData(data []byte, dataBlocks, parityBlocks int) [][]byte {
+	encodedData := mustEncodeData(data, dataBlocks, parityBlocks)
+	encodedData[7] = nil
+	encodedData[1] = nil
+	return encodedData
+}
+
+// Generates bad encoded data with one parity block and data block with garbage data.
+func genBadEncodedData(data []byte, dataBlocks, parityBlocks int) [][]byte {
+	encodedData := mustEncodeData(data, dataBlocks, parityBlocks)
+	encodedData[7] = []byte("garbage")
+	encodedData[1] = []byte("garbage")
+	return encodedData
+}
+
+// Generates encoded data with all data blocks missing.
+func genInsufficientData(data []byte, dataBlocks, parityBlocks int) [][]byte {
+	encodedData := mustEncodeData(data, dataBlocks, parityBlocks)
+	for i := 0; i < dataBlocks+1; i++ {
+		encodedData[i] = nil
+	}
+	return encodedData
+}
+
+// TestEncodeDecodeErasureData - tests erasure encoding and decoding.
+func TestEncodeDecodeErasureData(t *testing.T) {
+	data := []byte("Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.")
+	testEncodeCases := []struct {
+		inputData    []byte
+		dataBlocks   int
+		parityBlocks int
+		shouldPass   bool
+	}{
+		{
+			inputData:    data,
+			dataBlocks:   4,
+			parityBlocks: 4,
+			shouldPass:   true,
+		},
+		{
+			inputData:    []byte(""),
+			dataBlocks:   4,
+			parityBlocks: 4,
+			shouldPass:   false,
+		},
+	}
+	for i, testCase := range testEncodeCases {
+		_, err := encodeData(testCase.inputData, testCase.dataBlocks, testCase.parityBlocks)
+		if err != nil && testCase.shouldPass {
+			t.Errorf("Test %d: Expected to pass by failed instead with %s", i+1, err)
+		}
+	}
+
+	testDecodeCases := []struct {
+		encodedData  [][]byte
+		dataBlocks   int
+		parityBlocks int
+		blockSize    int
+		shouldPass   bool
+	}{
+		{
+			encodedData:  genBadEncodedData(data, 4, 4),
+			dataBlocks:   4,
+			parityBlocks: 4,
+			blockSize:    len(data),
+			shouldPass:   false,
+		},
+		{
+			encodedData:  genGoodEncodedData(data, 4, 4),
+			dataBlocks:   4,
+			parityBlocks: 4,
+			blockSize:    len(data),
+			shouldPass:   true,
+		},
+		{
+			encodedData:  genInsufficientData(data, 4, 4),
+			dataBlocks:   4,
+			parityBlocks: 4,
+			blockSize:    len(data),
+			shouldPass:   false,
+		},
+	}
+
+	for i, testCase := range testDecodeCases {
+		err := decodeData(testCase.encodedData, testCase.dataBlocks, testCase.parityBlocks)
```

Comment:
Comments required here. We should also generalize these tests for 4x4, 6x6 and 8x8 configs.